net/socks5/server.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package socks5 is an auth-less SOCKS5 server implementation
+// for userspace networking in Tailscale.
+package socks5
+
+import (
+	"net"
+
+	"tailscale.com/types/logger"
+)
+
+// SOCKS5Version is the byte that represents the SOCKS version
+// in requests.
+const SOCKS5Version byte = 5
+
+// Command are the bytes sent in SOCKS5 packets
+// that represent the kind of connection the client needs.
+type Command byte
+
+// The set of valid SOCKS5 commans as described in RFC 1928.
+const (
+	Connect      Command = 1
+	Bind         Command = 2
+	UDPAssociate Command = 3
+)
+
+// Addr are the bytes sent in SOCKS5 packets
+// that represent particular address types.
+type Addr byte
+
+// The set of valid SOCKS5 address types as defined in RFC 1928.
+const (
+	IPv4       Addr = 1
+	DomainName Addr = 3
+	IPv6       Addr = 4
+)
+
+// ListenAndServe creates a SOCKS5 server at the given address:port.
+func ListenAndServe(address string, dialContext DialContext, logf logger.Logf) error {
```

Comment:
instead of this, do:

```go
type Server struct {
    // Logf optionally specifies the logger to use.
    // If nil, the standard logger is used
    Logf logger.Logf

    // Dialer optionally specifies the dialer to use for outgoing connections.
    // If nil, the net package's standard dialer is used.
    Dialer func(ctx context.Context, network, addr string) (net.Conn, error)
}

func (s *Sever) Serve(ln net.Listener) error {
    // ...
}
```

Basically, like https://golang.org/pkg/net/http/#Server.Serve.

Then make unexported methods on Server that return non-nil dialer and logf funcs.

The caller can do the net.Listen and pass it in.